components/script/dom/bindings/str.rs

```diff
@@ -113,6 +113,63 @@ pub fn is_token(s: &[u8]) -> bool {
     })
 }
 
+/// Returns whether the language is matched, as defined by
+/// [RFC 4647](https://tools.ietf.org/html/rfc4647#section-3.3.2).
+pub fn extended_filtering(tag: &str, range: &str) -> bool {
+    let lang_ranges: Vec<&str> = range.split(',').collect();
+
+    lang_ranges.iter().any(|&lang_range| {
+        // step 1
+        let range_subtags: Vec<&str> = lang_range.split('\x2d').collect();
+        let tag_subtags: Vec<&str> = tag.split('\x2d').collect();
+        if tag.eq_ignore_ascii_case(lang_range) || lang_range.eq_ignore_ascii_case("*") {
+            return true;
+        }
+
+        let mut range_iter = range_subtags.iter();
+        let mut tag_iter = tag_subtags.iter();
+
+        // step 2
+        // Note: [Level-4 spec](https://drafts.csswg.org/selectors/#lang-pseudo) check for wild card
+        if let (Some(range_subtag), Some(tag_subtag)) = (range_iter.next(), tag_iter.next()) {
+            if !(range_subtag.eq_ignore_ascii_case(tag_subtag) || range_subtag.eq_ignore_ascii_case("*")) {
+                return false;
+            }
+        }
+
+        // step 3
+        for range_subtag in range_iter {
+            // step 3a
+            if range_subtag.eq_ignore_ascii_case("*") {
+                continue;
+            }
+            loop {
```

Comment:
First, I do believe that this loop has to go away, regardless of which step we take next, because step 3e continues this loop, rather than the outer for loop.
So, 2 possible solutions here, either we duplicate substep A and B after the for loop, or we follow the spec strictly and set up some variables first before the for loop, so something like `let mut current_tag_subtag = tag_iter.next();` and we match against current_tag_subtag and updating it as necessary as outlined in the spec.